lib/Sema/CSSimplify.cpp

```diff
@@ -13595,21 +13595,43 @@ ConstraintSystem::simplifyExplicitGenericArgumentsConstraint(
     }
 
     decl = overloadChoice.getDecl();
+
     auto openedOverloadTypes = getOpenedTypes(overloadLocator);
     openedTypes.append(openedOverloadTypes.begin(), openedOverloadTypes.end());
   }
 
-  auto genericContext = decl->getAsGenericContext();
-  if (!genericContext)
+  std::function<GenericParamList *(ValueDecl *)> getGenericParams =
+      [&](ValueDecl *decl) -> GenericParamList * {
+    auto genericContext = decl->getAsGenericContext();
+    if (!genericContext)
+      return nullptr;
+
+    auto genericParams = genericContext->getGenericParams();
+    if (!genericParams || genericParams->size() == 0) {
+      // If declaration is a non-generic typealias, let's point
+      // to the underlying generic declaration.
+      if (auto *TA = dyn_cast<TypeAliasDecl>(decl)) {
+        if (TA->isGeneric())
+          return nullptr;
+        if (auto underlying = TA->getUnderlyingType()->getAnyNominal())
```

Comment:
How about checking if `TA->getUnderlyingType()-is<UnboundGenericType>()`?

Reply:
Other places I saw check for isGeneric but I can change if that’s better

Reply:
@slavapestov I made it `AnyGenericType` to make it possible to properly diagnose cases like `typealias Print = Printer<Any>`, using `UnboundGenericType` won't work because we'd just return `nullptr` and following logic would just return `SolutionKind::Solved` and I don't want to mess with that in this PR.